pilot/pkg/serviceregistry/kube/controller/ambientindex_workloadentry_test.go

```diff
@@ -0,0 +1,438 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"net/netip"
+	"strings"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	istiov1alpha3 "istio.io/api/networking/v1alpha3"
+	authz "istio.io/api/security/v1beta1"
+	"istio.io/api/type/v1beta1"
+	apiv1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/util/xdsfake"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/kube/kclient/clienttest"
+	"istio.io/istio/pkg/test"
+	"istio.io/istio/pkg/test/util/assert"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+func TestAmbientIndex_WorkloadEntries(t *testing.T) {
+	test.SetForTest(t, &features.EnableAmbientControllers, true)
+	cfg := memory.NewSyncController(memory.MakeSkipValidation(collections.PilotGatewayAPI()))
+	controller, fx := NewFakeControllerWithOptions(t, FakeControllerOptions{
+		ConfigController:     cfg,
+		MeshWatcher:          mesh.NewFixedWatcher(&meshconfig.MeshConfig{RootNamespace: "istio-system"}),
+		ClusterID:            "cluster0",
+		WorkloadEntryEnabled: true,
+	})
+	pc := clienttest.Wrap(t, controller.podsClient)
+	sc := clienttest.Wrap(t, controller.services)
+	cfg.RegisterEventHandler(gvk.AuthorizationPolicy, controller.AuthorizationPolicyHandler)
+	cfg.RegisterEventHandler(gvk.WorkloadEntry, controller.WorkloadEntryHandler)
+	go cfg.Run(test.NewStop(t))
+	addPolicy := func(name, ns string, selector map[string]string) {
```

Comment:
I think most of these functions are exact copies of the ones in `ambientindex_test.go`